chore(golangci-lint): add stylecheck linter

matgomes21 opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Enable stylecheck linter.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2030 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

mboersma commented:
/ok-to-test

sonasingh46 commented:
Hey @matgomes21 -- We need to silent the linters for couple of files around apis and conversions. I will post more details here after verifying.

Reason: Some files are autogenerated and have their own style. Now there are some files that are not autogenerated but may need to implement some methods from autogenerated ones. So we want to be consistent and silent the linter for these files.

We will need to put some regular expression that selects some path in the repo for which the linter should ignore. 
Something like this: 
https://github.com/kubernetes-sigs/cluster-api/blob/main/.golangci.yml#L228

matgomes21 commented:
Hey @sonasingh46 , I understood the issue, so I will wait for more details of which files should be ignored by linter.

mboersma commented:
@matgomes21, thanks for the PR! 

It looks like the linter-related changes are fine (`make lint` passes for me on this branch).  But changing the case of the string in the error message `exp/api/v1beta1/azuremachinepool_webhook.go` actually broke a unit test expectation in `azuremachinepool_test.go`.  If you fix the strings in the test to match and make sure `make test` passes, I think this is good.

matgomes21 commented:
@mboersma, thanks for the review! I will work on it right away.

matgomes21 commented:
Hey @mboersma, I fixed the tests with string problems and `make test` passed. Can you review again the PR? Thanks!

mboersma reviewed (approved):
/lgtm

CecileRobertMichon commented on exp/api/v1beta1/azuremanagedmachinepool_types.go:
```diff
@@ -165,23 +165,23 @@ type AzureManagedMachinePoolList struct {
 // GetConditions returns the list of conditions for an AzureManagedMachinePool API object.
-func (m *AzureManagedMachinePool) GetConditions() clusterv1.Conditions {
-	return m.Status.Conditions
+func (r *AzureManagedMachinePool) GetConditions() clusterv1.Conditions {
```
why "r" as a receiver here?

sonasingh46 replied:
For `AzureManagedMachinePool ` -- code base has `r` as well as `m` as receivers and the `stylecheck` linter complains about that as consistency is preferred. Semantically, `m` looks more appropriate to me as compared to `r`

matgomes21 replied:
I opted for using `r` over `m` because it had more ocurrencies, but I agree thar `m` looks more appropriate, I will make the change.

sonasingh46 commented on api/v1alpha3/azureclusteridentity_conversion.go:
```diff
@@ -30,25 +30,25 @@ const (
+func (c *AzureClusterIdentity) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*infrav1beta1.AzureClusterIdentity)
-	if err := Convert_v1alpha3_AzureClusterIdentity_To_v1beta1_AzureClusterIdentity(src, dst, nil); err != nil {
+	if err := Convert_v1alpha3_AzureClusterIdentity_To_v1beta1_AzureClusterIdentity(c, dst, nil); err != nil {
```
these files receiver names should not be changed and ignored.

sonasingh46 replied:
We can add following in the golangci lint yaml 
```
    - linters:
        - stylecheck
      text: "ST1016: methods on the same type should have the same receiver name"
      path: .*(api|types)\/.*\/*_types\.go$
```

sonasingh46 replied:
Add above after this line : 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/.golangci.yml#L102

matgomes21 commented:
Hey @sonasingh46 @mboersma @CecileRobertMichon , I changed the `r` receivers to `m` and added the regular expression to silence linter on api files. If there are any other change suggestions, let me know!

mboersma commented:
/retest

mboersma commented:
/retest

mboersma reviewed (approved):
/lgtm

CecileRobertMichon commented:
/lgtm
/approve
